diff -urN -x '.files-list-*' -x '.applied*' -x .applied_patches_list -x .br_regen_dot_config -x '*include-prefixes*' linux-6.19.4/arch/arm/boot/dts/st/stm32f429-disco.dts ../buildroot-2026.02/output/build/linux-6.19.4/arch/arm/boot/dts/st/stm32f429-disco.dts
--- linux-6.19.4/arch/arm/boot/dts/st/stm32f429-disco.dts	2026-02-27 07:01:41.000000000 +0800
+++ ../buildroot-2026.02/output/build/linux-6.19.4/arch/arm/boot/dts/st/stm32f429-disco.dts	2026-04-16 17:39:18.093484211 +0800
@@ -57,7 +57,7 @@
 	compatible = "st,stm32f429i-disco", "st,stm32f429";
 
 	chosen {
-		bootargs = "root=/dev/ram";
+		bootargs = "root=mtd:romfs rootfstype=romfs ro";
 		stdout-path = "serial0:115200n8";
 	};
 
@@ -66,6 +66,18 @@
 		reg = <0x90000000 0x800000>;
 	};
 
+    	flash@8000000 {
+		compatible = "mtd-rom";
+		reg = <0x8000000 0x200000>;
+		bank-width = <4>;
+		#address-cells = <1>;
+
+		partition@19d000 {
+			label = "romfs";
+			reg = <0x19d000 0x63000>;
+		};
+	};
+
 	aliases {
 		serial0 = &usart1;
 	};
@@ -98,6 +110,14 @@
 		regulator-name = "vcc5_host1";
 		regulator-always-on;
 	};
+
+    v3v3: regulator-v3v3 {
+        compatible = "regulator-fixed";
+        regulator-name = "v3v3";
+        regulator-min-microvolt = <3300000>;
+        regulator-max-microvolt = <3300000>;
+        regulator-always-on;  
+    };
 };
 
 &clk_hse {
@@ -230,3 +250,14 @@
 	pinctrl-names = "default";
 	status = "okay";
 };
+
+&sdio {
+    status = "okay";
+    pinctrl-names = "default", "opendrain";
+    pinctrl-0 = <&sdio_pins>;
+    pinctrl-1 = <&sdio_pins_od>;
+    bus-width = <1>;
+    max-frequency = <1000000>;
+    broken-cd;
+    vmmc-supply = <&v3v3>;
+};
diff -urN -x '.files-list-*' -x '.applied*' -x .applied_patches_list -x .br_regen_dot_config -x '*include-prefixes*' linux-6.19.4/drivers/char/hw_random/stm32-rng.c ../buildroot-2026.02/output/build/linux-6.19.4/drivers/char/hw_random/stm32-rng.c
--- linux-6.19.4/drivers/char/hw_random/stm32-rng.c	2026-02-27 07:01:41.000000000 +0800
+++ ../buildroot-2026.02/output/build/linux-6.19.4/drivers/char/hw_random/stm32-rng.c	2026-04-16 17:39:18.099484276 +0800
@@ -363,11 +363,21 @@
		return -EINVAL;
	}

-	clk_bulk_disable_unprepare(priv->data->nb_clock, priv->clk_bulk);
+	if (IS_ENABLED(CONFIG_PM))
+		clk_bulk_disable_unprepare(priv->data->nb_clock, priv->clk_bulk);

	return 0;
 }

+static void stm32_rng_cleanup(struct hwrng *rng)
+{
+		struct stm32_rng_private *priv =
+			container_of(rng, struct stm32_rng_private, rng);
+		if (!IS_ENABLED(CONFIG_PM))
+		clk_bulk_disable_unprepare(priv->data->nb_clock, priv->clk_bulk);
+}
+
+
 static void stm32_rng_remove(struct platform_device *ofdev)
 {
	pm_runtime_disable(&ofdev->dev);
@@ -563,6 +573,7 @@

	priv->rng.name = dev_driver_string(dev);
	priv->rng.init = stm32_rng_init;
+    priv->rng.cleanup = stm32_rng_cleanup;
	priv->rng.read = stm32_rng_read;
	priv->rng.quality = 900;

diff -urN -x '.files-list-*' -x '.applied*' -x .applied_patches_list -x .br_regen_dot_config -x '*include-prefixes*' linux-6.19.4/drivers/gpu/drm/stm/ltdc.c ../buildroot-2026.02/output/build/linux-6.19.4/drivers/gpu/drm/stm/ltdc.c
--- linux-6.19.4/drivers/gpu/drm/stm/ltdc.c	2026-02-27 07:01:41.000000000 +0800
+++ ../buildroot-2026.02/output/build/linux-6.19.4/drivers/gpu/drm/stm/ltdc.c	2026-04-16 17:39:18.104484330 +0800
@@ -2079,16 +2079,16 @@
 		drm_err(ddev, "Failed calling drm_vblank_init()\n");
 		goto err;
 	}
+    if (IS_ENABLED(CONFIG_PM)) {
+        clk_disable_unprepare(ldev->pixel_clk);
 
-	clk_disable_unprepare(ldev->pixel_clk);
-
-	if (ldev->bus_clk)
-		clk_disable_unprepare(ldev->bus_clk);
-
-	pinctrl_pm_select_sleep_state(ddev->dev);
+        if (ldev->bus_clk)
+            clk_disable_unprepare(ldev->bus_clk);
 
-	pm_runtime_enable(ddev->dev);
+        pinctrl_pm_select_sleep_state(ddev->dev);
 
+        pm_runtime_enable(ddev->dev);
+    }
 	return 0;
 err:
 	clk_disable_unprepare(ldev->pixel_clk);
@@ -2101,9 +2101,16 @@
 
 void ltdc_unload(struct drm_device *ddev)
 {
+    struct ltdc_device *ldev = ddev->dev_private;
 	drm_dbg_driver(ddev, "\n");
 
-	pm_runtime_disable(ddev->dev);
+    if (IS_ENABLED(CONFIG_PM))
+	    pm_runtime_disable(ddev->dev);
+    else {
+        clk_disable_unprepare(ldev->pixel_clk);
+        if (ldev->bus_clk)
+            clk_disable_unprepare(ldev->bus_clk);
+    }
 }
 
 MODULE_AUTHOR("Philippe Cornu <philippe.cornu@st.com>");
