--- package/uclibc/uclibc.mk	2026-03-04 16:11:17.000000000 +0800
+++ ../../buildroot-2026.02-rc2/package/uclibc/uclibc.mk	2026-03-15 21:27:25.682220959 +0800
@@ -382,7 +382,7 @@
 # Commands
 #
 
-UCLIBC_EXTRA_CFLAGS = $(TARGET_ABI) $(TARGET_DEBUGGING)
+UCLIBC_EXTRA_CFLAGS = $(TARGET_ABI) $(TARGET_DEBUGGING) -fpic -mpic-register=r10 -Os -mthumb -mcpu=cortex-m4 -mno-pic-data-is-text-relative
 
 # uClibc-ng does not build with LTO, so explicitly disable it
 # when using a compiler that may have support for LTO
