From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cortexm-linux PR#19 <jserv@ccns.ncku.edu.tw>
Subject: [PATCH] musl fdpic dlstart direct dls2

---
diff --git a/ldso/dlstart.c b/ldso/dlstart.c
--- a/ldso/dlstart.c	2026-05-10 03:00:12.907595304 +0800
+++ b/ldso/dlstart.c	2026-05-10 03:00:12.908054428 +0800
@@ -15,6 +15,10 @@
 	__asm__ __volatile__ ( "" : "+m"(static_func_ptr) : : "memory"); \
 	*(fp) = static_func_ptr; } while(0)
 #endif
+
+#if DL_FDPIC
+hidden void __dls2(unsigned char *, size_t *);
+#endif
 
 hidden void _dlstart_c(size_t *sp, size_t *dynv)
 {
@@ -154,7 +158,11 @@
 	}
 #endif
 
+#if DL_FDPIC
+	__dls2((void *)base, sp);
+#else
 	stage2_func dls2;
 	GETFUNCSYM(&dls2, __dls2, base+dyn[DT_PLTGOT]);
 	dls2((void *)base, sp);
+#endif
 }
