From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cortexm-linux PR#19 <jserv@ccns.ncku.edu.tw>
Subject: [PATCH] musl fdpic local funcdesc got

---
diff --git a/ldso/dynlink.c b/ldso/dynlink.c
--- a/ldso/dynlink.c	2026-05-10 03:00:12.907595304 +0800
+++ b/ldso/dynlink.c	2026-05-10 03:00:12.908054428 +0800
@@ -515,7 +515,10 @@
 		case REL_FUNCDESC_VAL:
 			if ((sym->st_info&0xf) == STT_SECTION) *reloc_addr += sym_val;
 			else *reloc_addr = sym_val;
-			reloc_addr[1] = def.sym ? (size_t)def.dso->got : 0;
+			if ((sym->st_info&0xf) == STT_SECTION)
+				reloc_addr[1] = (size_t)dso->got;
+			else
+				reloc_addr[1] = def.sym ? (size_t)def.dso->got : 0;
 			break;
 		case REL_DTPMOD:
 			*reloc_addr = def.dso->tls_id;
